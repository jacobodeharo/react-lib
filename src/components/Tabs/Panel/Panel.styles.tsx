import styled from 'styled-components';

const StyledPanel = styled.div`
  background-color: var(--c-panel-bg, ${({ theme }) => theme.primaryColorLight});
  color: var(--c-panel-fg, ${({ theme }) => theme.primaryColorText});
  padding: 10px;
  border-radius: 0 0 5px 5px;
`;

export { StyledPanel };
