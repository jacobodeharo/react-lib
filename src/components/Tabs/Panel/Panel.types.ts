import { Theme } from '@/components/theme';

export type PanelProps = {
  tabId: string;
  theme?: Theme;
};
