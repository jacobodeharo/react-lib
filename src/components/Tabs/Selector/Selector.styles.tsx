import styled from 'styled-components';

const StyledSelector = styled.div`
  align-items: center;
  background: gold;
  display: flex;
  flex: 1 1 auto;
  justify-content: center;
  min-height: 35px;
  margin: 10px;

  :hover {
    cursor: pointer;
    background: darkgoldenrod;
  }
`;

export { StyledSelector };
