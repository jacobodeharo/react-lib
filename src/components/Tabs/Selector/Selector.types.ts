import { Theme } from '@/components/theme';

export type SelectorProps = {
  tabId: string;
  active?: boolean;
  hovered?: boolean;
  theme?: Theme;
};
