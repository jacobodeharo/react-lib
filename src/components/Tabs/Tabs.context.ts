import { TTabsContext } from './Tabs.types';
import * as React from 'react';

const TabsContext = React.createContext<TTabsContext | undefined>(undefined);

export { TabsContext };
