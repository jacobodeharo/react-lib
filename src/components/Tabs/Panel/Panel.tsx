import React from 'react';
import { PanelProps } from './Panel.types';
import { useTabs } from '../Tabs.hooks';

const Panel: React.FC<PanelProps> = (props) => {
  const { activeTab } = useTabs();
  const { children, tabId } = props;
  return activeTab === tabId ? <div>{children}</div> : null;
};

export { Panel };
