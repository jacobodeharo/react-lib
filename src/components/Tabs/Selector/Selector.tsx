import React from 'react';
import { SelectorProps } from './Selector.types';
import { useTabs } from '../Tabs.hooks';
import { StyledSelector } from './Selector.styles';

const Selector: React.FC<SelectorProps> = (props) => {
  const { setActiveTab } = useTabs();
  const { tabId, children } = props;
  const clickHandler = () => {
    setActiveTab(tabId);
  };
  return (
    <StyledSelector onClick={clickHandler} {...props}>
      {children}
    </StyledSelector>
  );
};

export { Selector };
