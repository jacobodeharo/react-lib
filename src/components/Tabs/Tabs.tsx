import * as React from 'react';
import { TabsProps } from './Tabs.types';
import { TabsContext } from './Tabs.context';
import { Selector } from './Selector/Selector';
import { Panel } from './Panel/Panel';
import { renderChildrenOfType } from '../../helper/reactDomUtils';
import { TabsBody, TabsHeader, TabsWrapper } from './Tabs.styles';

const Tabs: React.FC<TabsProps> = (props) => {
  const { children, selectedTab } = props;
  const [activeTab, setActiveTab] = React.useState(selectedTab || '');
  const memoizedContextValue = React.useMemo(
    () => ({
      activeTab,
      setActiveTab,
    }),
    [activeTab, setActiveTab]
  );
  return (
    <TabsContext.Provider value={memoizedContextValue}>
      <TabsWrapper>
        <TabsHeader>{renderChildrenOfType(children, Selector)}</TabsHeader>
        <TabsBody>{renderChildrenOfType(children, Panel)}</TabsBody>
      </TabsWrapper>
    </TabsContext.Provider>
  );
};

export { Tabs };
