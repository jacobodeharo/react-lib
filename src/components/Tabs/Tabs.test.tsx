import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import { Tabs } from './Tabs';
import { AMBER_THEME } from '../theme/variants/amberTheme';
import { Selector } from './Selector/Selector';
import { Panel } from './Panel/Panel';
import { useTabs } from './Tabs.hooks';
import { TabsContext } from './Tabs.context';

describe('Tabs', () => {
  const PANEL_CHILDREN = 'Panel Label';
  const SELECTOR_CHILDREN = 'SelectorLabel';
  afterEach(cleanup);

  test('has to be rendered', () => {
    render(
      <Tabs selectedTab='tabA'>
        <Selector tabId='tabA'>{SELECTOR_CHILDREN}</Selector>
        <Panel tabId='tabA'>{PANEL_CHILDREN}</Panel>
      </Tabs>
    );
    const PanelElement = screen.getByText(PANEL_CHILDREN);
    expect(PanelElement).toBeInTheDocument();
    const SelectorElement = screen.getByText(SELECTOR_CHILDREN);
    expect(SelectorElement).toBeInTheDocument();
    expect(SelectorElement).toMatchSnapshot();
  });

  test('has to accept theming', () => {
    render(
      <Tabs selectedTab='tabA' theme={AMBER_THEME}>
        <Selector tabId='tabA'>{SELECTOR_CHILDREN}</Selector>
        <Panel tabId='tabA'>{PANEL_CHILDREN}</Panel>
      </Tabs>
    );
    const PanelElement = screen.getByText(SELECTOR_CHILDREN);
    expect(PanelElement).toBeInTheDocument();
    expect(PanelElement).toMatchSnapshot();
  });
});

describe('useTabs hook', () => {
  test('has to create a TabsContext', () => {
    const setActiveTabMock = () => undefined;
    const context = renderHook(() => useTabs(), {
      initialProps: {
        value: { activeTab: 'TabA', setActiveTab: setActiveTabMock },
      },
      wrapper: TabsContext.Provider,
    });
    expect(context.result).toStrictEqual({
      current: {
        activeTab: 'TabA',
        setActiveTab: setActiveTabMock,
      },
      error: undefined,
    });
  });
});
