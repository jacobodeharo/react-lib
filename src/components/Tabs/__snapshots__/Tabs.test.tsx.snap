// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs has to accept theming 1`] = `
<div
  class="sc-bdfBwQ iRGlIU"
>
  SelectorLabel
</div>
`;

exports[`Tabs has to be rendered 1`] = `
<div
  class="sc-bdfBwQ iRGlIU"
>
  SelectorLabel
</div>
`;
